src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts

```diff
@@ -116,4 +136,15 @@ export class NavigationComponent implements OnInit, OnDestroy {
       }
     }
   }
+
+  private restoreDisplayedSubMenu(): void {
+    const currentPath = window?.location?.hash?.split?.('/')?.[1];
+    if (!currentPath) return;
+    for (const [group, paths] of Object.entries(this.groupedPaths)) {
+      if (paths.includes(currentPath)) {
+        this.toggleSubMenu(group);
+        break;
+      }
+    }
+  }
```

Comment:
Is there any way to achieve this but using the [`RouterLinkActive`](https://angular.io/api/router/RouterLinkActive) directive instead of adding extra logic (I'd also remove the `toggleSubMenu()` method and replace that with Angular template stuff).

`RouterLinkActive` already detects the activated route and allows us to set a CSS class or an ARIA attribute based on that (Patternfly styles expanded menus based on `aria-expanded`, while [IBM Carbon](https://github.com/carbon-design-system/carbon-components-angular/blob/master/src/ui-shell/sidenav/sidenav.component.ts) used both `aria-expanded` and `aria-current`).

So, what about something like [this](https://stackoverflow.com/questions/52884264/angular-6-how-do-i-set-a-parent-menu-item-to-active-using-routerlinkactive-whe)?

Reply:
Hi @epuertat 

To achieve something like this, I would need to group the routes based on the html grouping in the navigation.html. Some discussion would be necessary because some things don't make sense. For example, in the left navigation menu,  the route ```cluster``` is grouped under  "Administration". ```nfs``` is grouped under both "Object" and "File"!

Also, there are paths that won't match any of the links in the left navigation menu like ```perf_counters``` for example, so relying only on ```routerLinkActive``` won't work unless this route is grouped. 

So only way to achieve this is to group the routes in ```app-routing.module.ts```.

Reply:
> To achieve something like this, I would need to group the routes based on the html grouping in the navigation.html. 

That's a great idea! IMHO the URL hierarchy, breadcrumbs and site navigation should all be consistent with one another, so:
* We would need to update and nest the routes, so that components under the `Cluster >` menu are available at `/cluster/{component}` URL as well:
  ```ts
  const routes: Routes = [
    ...
    children: [
      ...  
      // Cluster
      {
        path: 'cluster',
        children: [
          {
            path: 'hosts',
            component: HostsComponent,
            data: { breadcrumbs: 'Cluster/Hosts' },
  ```
* We'd need to update the links at `navigation.component.html`.
* [Optional clean-up] Ensure that the code paths also follow that convention (e.g.: I see that most Cluster menu sub-items are under `.../frontend/src/app/ceph/cluster` dir.
* [Optional clean-up] Instead of loading all routes in the main `app-routing.module.ts`, we could define those routes in each parent menu module (for example, we already [did that for RGW](https://github.com/ivoalmeida/ceph/blob/dfe8ceb5c59afaffd534469613520ac584b54f06/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw.module.ts#L110-L206)), so something similar would be great for the [cluster parent menu](https://github.com/ivoalmeida/ceph/blob/restore-menu-state/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/cluster.module.ts).  
* [Optionally] if we want to ensure a smooth transition (to avoid broken UI links, if any), we can keep the old URLs still available but redirecting to the new nested routes:
  ```ts
  { path: 'hosts', redirectTo: '/cluster/hosts'},
  ``` 

> Some discussion would be necessary because some things don't make sense. For example, in the left navigation menu, the route `cluster` is grouped under "Administration". `nfs` is grouped under both "Object" and "File"!

I don't think it's a big issue: we can keep the same component (e.g.: NFS) available a different URLs (`/rgw/nfs` and `/cephfs/nfs`) and parent menus (`Object > NFS` and `File > NFS`). As a follow-up PR to this improvement, what we can do instead is customize that component based on the parent menu. For example, `/rgw/nfs` only shows RGW-backed NFS mount points, while `/cephfs/nfs` would do the same for CephFS-backed mount points.

> Also, there are paths that won't match any of the links in the left navigation menu like `perf_counters` for example, so relying only on `routerLinkActive` won't work unless this route is grouped.

`RouterLink` performs by default a partial match, so putting everything under `/cluster/{component}` would work, as long as we move those components to a route under `cluster`.

> So only way to achieve this is to group the routes in `app-routing.module.ts`.

Yes, that'd be the best way.

If we don't want that approach, another approach that still reduces the duplication is to use the templates variables from the `RouterLinkActive` directives, something like this:
```html
<!-- Cluster -->
<ul>
  <li>
    <a [attr.aria-expanded]="poolsLink.isActive || hostsLink.isActive || ..." >
      Cluster
    </a>
    <ul>
      <li><a routerLink="/pools" routerLinkActive #poolsLink="routerLinkActive">Pools</a></li>
      <li><a routerLink="/hosts" routerLinkActive #hostsLink="routerLinkActive">Hosts</a></li>
      ...
    </ul>
  </li>
```

Reply:
yw @ivoalmeida, glad to be useful!

UI routes are not as strict as back-end routes (which we should keep stable, not only for the Dashboard front-end, but for potential RESTful users as well), so we are free to move them (@nizamial09 just to confirm: is IBM's SI using hardcoded URLs to the Dashboard?).

I just saw that CRUD (Create-Read-Update-Delete) children routes [are rather redundant](https://github.com/ivoalmeida/ceph/blob/dfe8ceb5c59afaffd534469613520ac584b54f06/src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts#L123-L154), so we could create a factory method to simplify those.

Instead of (which IMHO is incorrect, since `ceph-users` should rather be `/cluster/user` or `/admin/user`):
```ts
      {
        path: 'ceph-users',
        component: CRUDTableComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/create',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Create',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/import',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Import',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/edit',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Edit',
          resource: 'api.cluster.user@1.0'
        }
      },
```
Something like (the `breadcrumb` service should be smart enough to concatenate parent breadcrumbs):
```ts
    createCRUDRoutes({
      title: $localize`Ceph Users`,
      basePath: `admin/user`,
      actions: [Action.INDEX, Action.CREATE, Action.EDIT, Action.IMPORT],
      components: {
        Action.INDEX: CRUDTableComponent,
        Action.DEFAULT: CrudFormComponent
      },
      data: {
        resource: 'api.cluster.user@1.0'
      }
    })
```

Reply:
yw @ivoalmeida, glad to be useful!

UI routes are not as strict as back-end routes (which we should keep stable, not only for the Dashboard front-end, but for potential RESTful users as well), so we are free to move them (@nizamial09 just to confirm: is IBM's SI using hardcoded URLs to the Dashboard?).

I just saw that CRUD (Create-Read-Update-Delete) children routes [are rather redundant](https://github.com/ivoalmeida/ceph/blob/dfe8ceb5c59afaffd534469613520ac584b54f06/src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts#L123-L154), so we could create a factory method to simplify those.

Instead of (which IMHO is incorrect, since `ceph-users` should rather be `/cluster/user` or `/admin/user`):
```ts
      {
        path: 'ceph-users',
        component: CRUDTableComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/create',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Create',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/import',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Import',
          resource: 'api.cluster.user@1.0'
        }
      },
      {
        path: 'cluster/user/edit',
        component: CrudFormComponent,
        data: {
          breadcrumbs: 'Administration/Ceph Users/Edit',
          resource: 'api.cluster.user@1.0'
        }
      },
```
Something like (the `breadcrumb` service should be smart enough to concatenate parent breadcrumbs):
```ts
    createCRUDRoutes({
      title: $localize`Ceph Users`,
      basePath: `admin/user`,
      actions: [Action.INDEX, Action.CREATE, Action.EDIT, Action.IMPORT],
      components: {
        Action.INDEX: CRUDTableComponent,
        Action.DEFAULT: CrudFormComponent
      },
      data: {
        resource: 'api.cluster.user@1.0'
      }
    })
```

Reply:
BTW, in the specific case of the `CRUD*Component` the idea was that all this metadata was dynamically generated from the back-end, in order to minimize redundancies in back-end & front-end layer, and make the front-end just a wrapper of the back-end data structures (Python dictionaries/REST collections becoming Angular/Bootstrap tables, and REST actions resulting in Angular Forms). We can talk of this in more detail in our meeting ;-)

Reply:
> If we don't want that approach, another approach that still reduces the duplication is to use the templates variables from the `RouterLinkActive` directives, something like this:
> 
> ```
> <!-- Cluster -->
> <ul>
>   <li>
>     <a [attr.aria-expanded]="poolsLink.isActive || hostsLink.isActive || ..." >
>       Cluster
>     </a>
>     <ul>
>       <li><a routerLink="/pools" routerLinkActive #poolsLink="routerLinkActive">Pools</a></li>
>       <li><a routerLink="/hosts" routerLinkActive #hostsLink="routerLinkActive">Hosts</a></li>
>       ...
>     </ul>
>   </li>
> ```
> 
> What do you think of this?

this doesn't work. you are trying to access a reference to a child element from a parent.